src/main/java/mx/erickb/shipping/util/route/RouteUtils.java

```diff
@@ -0,0 +1,77 @@
+package mx.erickb.shipping.util.route;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import mx.erickb.shipping.exception.InvalidResponseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+
+@Component
+public class RouteUtils {
+
+    private final Logger logger = LoggerFactory.getLogger(RouteUtils.class);
+    private final ObjectMapper mapper;
+
+    public RouteUtils(ObjectMapper mapper) {
+        this.mapper = mapper;
+    }
+
+    public List<Route> parseRoutes(String routesResponse) throws InvalidResponseException {
+        try {
+            Route[] routes = mapper.readValue(routesResponse, Route[].class);
+            return Arrays.asList(routes);
+        } catch (JsonProcessingException e) {
+            logger.error("Error when processing server response (" + e.getMessage() + ")");
+            throw new InvalidResponseException("Invalid response: \"" + routesResponse + '"');
+        }
+    }
+
+    public RoutesGraph getRoutesGraph(List<Route> routes) {
+        // Count unique cities
+        Set<String> citiesSet = new HashSet<>();
+        for (Route route : routes) {
+            citiesSet.add(route.getFrom());
+            citiesSet.add(route.getTo());
+        }
+
+        List<String> cities = new ArrayList<>(citiesSet);
+        RoutesGraph routesGraph = new RoutesGraph(cities);
+        // Filling connections
+        for (Route route : routes) {
+            routesGraph.addRoute(route.getFrom(), route.getTo(), route.getDistance());
+        }
+        return routesGraph;
+    }
+
+    public List<String> findOptimalRoute(RoutesGraph routesGraph, String origin, String destination) {
+        return this.findOptimalRoute(new ArrayList<>(), routesGraph, origin, destination);
+    }
+
+    // Depth first search (recursive) to find route (weights are discarded)
+    private List<String> findOptimalRoute(List<String> currentRoute, RoutesGraph routesGraph, String currentCity, String destination) {
+        // Clone current route to avoid interfering with recursion
+        List<String> route = new ArrayList<>(currentRoute);
+        route.add(currentCity);
+        if (currentCity.equals(destination)) {
+            return route;
+        }
+
+        // Recursively visit each connected city
+        for (String connectedCity : routesGraph.getConnectedCities(currentCity)) {
+            if (route.contains(connectedCity)) {
+                // Skip if already visited in current route (avoid cycles)
+                continue;
+            }
+            List<String> result = findOptimalRoute(route, routesGraph, connectedCity, destination);
+            if (result != null) {
+                return result;
+            }
```

Comment:
Si result es es `null` entonces continuara recorriendo el `foreach` y checará que todos los elementos ya estan contenidos en `route` no? pero ese es procesamiento innecesario, podrías hacer un `break` si `result == null`

Reply:
De hecho como mi intención es recorrer todo el grafo en profundidad, en este for estoy iterando sobre las ciudades que están conectadas a mi nodo actual, si no encuentro (recursivamente) un camino a mi ciudad destino por medio de una ciudad, aún tengo que revisar los demás caminos a partir de las demás ciudades conectadas y si me salgo del for usando un break no buscaré exhaustivamente el grafo y solo revisaría una ruta.
En cualquier caso no creo que convenga detenerme mucho en esta parte ya que en la historia SHP-07-2 se implementará un algoritmo distinto

Reply:
ok, esq como en este caso estas solo regresando la primera ruta que encuentras creo que sí podría hacer break, pero como dices en la siguiente historia tendrá que cambiar probablemente, por lo tanto esta bien así mientras.